stdlib/string.gr

```diff
@@ -519,3 +519,654 @@ export let endsWith = (p: String, s: String) => {
+        while (bytesPtr < end) {
+          let w1 = (WasmI32.load8U(bytesPtr, 0n) << 8n) | (WasmI32.load8U(bytesPtr, 1n))
+          let codeWord = if (w1 >= 0xD800n && w1 <= 0xDBFFn) {
+            // high surrogate. next character is low srurrogate
+            let w1 = (w1 & 0x03FFn) << 10n
+            let w2 = ((WasmI32.load8U(bytesPtr, 2n) << 8n) | (WasmI32.load8U(bytesPtr, 3n))) & 0x03FFn
+            let codeWord = w1 + w2 + 0x10000n
+            // no problems, so go past both code words
+            bytesPtr += 4n
+            codeWord
+          } else {
+            bytesPtr += 2n
+            w1
+          }
+          strPtr += writeUtf8CodePoint(strPtr, codeWord)
+        }
+      },
+      UTF16_LE => {
+        // NOTE: Because the size check passed, we know the string is well-formed
+        let end = bytesPtr + bytesSize
+        while (bytesPtr < end) {
+          let w1 = (WasmI32.load8U(bytesPtr, 1n) << 8n) | (WasmI32.load8U(bytesPtr, 0n))
+          let codeWord = if (w1 >= 0xD800n && w1 <= 0xDBFFn) {
+            // high surrogate. next character is low srurrogate
+            let w1 = (w1 & 0x03FFn) << 10n
+            let w2 = ((WasmI32.load8U(bytesPtr, 3n) << 8n) | (WasmI32.load8U(bytesPtr, 2n))) & 0x03FFn
+            //let uPrime = codePoint - 0x10000n
+            //let w1 = ((uPrime & 0b11111111110000000000n) >>> 10n) + 0xD800n // High surrogate
+            //let w2 = (uPrime & 0b00000000001111111111n) + 0xDC00n // Low surrogate
+            let codeWord = w1 + w2 + 0x10000n
+            // no problems, so go past both code words
+            bytesPtr += 4n
+            codeWord
+          } else {
+            bytesPtr += 2n
+            w1
+          }
+          strPtr += writeUtf8CodePoint(strPtr, codeWord)
+        }
+      },
+      UTF32_BE => {
+        let end = bytesPtr + bytesSize
+        while (bytesPtr < end) {
+          let codeWord = (WasmI32.load8U(bytesPtr, 0n) << 24n) | (WasmI32.load8U(bytesPtr, 1n) << 16n) | (WasmI32.load8U(bytesPtr, 2n) << 8n) | (WasmI32.load8U(bytesPtr, 3n))
+          bytesPtr += 4n
+          strPtr += writeUtf8CodePoint(strPtr, codeWord)
+        }
+      },
+      UTF32_LE => {
+        let end = bytesPtr + bytesSize
+        while (bytesPtr < end) {
+          let codeWord = (WasmI32.load8U(bytesPtr, 3n) << 24n) | (WasmI32.load8U(bytesPtr, 2n) << 16n) | (WasmI32.load8U(bytesPtr, 1n) << 8n) | (WasmI32.load8U(bytesPtr, 0n))
+          bytesPtr += 4n
+          strPtr += writeUtf8CodePoint(strPtr, codeWord)
+        }
+      }
+    }
+    WasmI32.toGrain(str) : String
+  }
+}
+
+// Decodes the given byte sequence into a string using the given encoding scheme
+// @param bytes: Bytes - The input bytes
+// @param encoding: Encoding - The encoding to use
+// @param skipBom: Bool - Whether to include the byte-order marker (if present) in the decoded output
+// @returns String
+@disableGC
+export let decode = (bytes: Bytes, encoding: Encoding, skipBom: Bool) => {
```

Comment:
Since we don't have optional arguments yet, I believe we should have separate methods for `withBOM`.

This would become `decode` and `decodeWithBOM` that passes that argument into the "help" function for `decodeRange`.